qemu/tests/uefishell.py

```diff
@@ -0,0 +1,147 @@
+                            env_process.preprocess_vm)
+
+        vm = self.env.get_vm(params["main_vm"])
+        self.session = vm.wait_for_serial_login()
+        if under_fs0 == "yes":
+            self.send_command("fs0:")
+
+    def var_copy(self):
+        """
+        Copy a new VAR file for a cleaner environment to test
+        """
+        source_var_path = "/usr/share/OVMF/OVMF_VARS.fd"
+        path = storage.get_image_filename_filesytem(
+            self.params, data_dir.get_data_dir())
+        vars_path = "%s.fd" % path
+        if os.path.exists(source_var_path):
+            logging.debug("Use new var file, copied from %s", source_var_path)
+            cmd = "cp -f %s %s" % (source_var_path, vars_path)
+            process.system(cmd)
+
+    def send_command(self, command, check_result=False):
+        """
+        Send a command line to uefi shell, and check the output
+        if 'check_result' exists, and fail the case if the output
+        does not meet the expectation
+
+        :param command: the command string being executed
+        :param check_result: the pattern to validate output
+        """
+        logging.info("Send uefishell command: %s" % command)
+        output = self.session.cmd_output_safe(command)
+        if check_result:
+            if not re.search(check_result, output):
+                self.test.fail("The command result is: %s, which does not"
+                               " match the expectation: %s" % (output, check_result))
+
+    def post_test(self):
+        """
+        To execute post test if have, e.g. guest installation. And
+        post test step should be defined within a function of this
+        class UEFIShell
+        """
+        post_tests = self.params.get("post_tests", "")
+        for test in post_tests.split():
+            if hasattr(self, test):
+                func = getattr(self, test)
+                func()
+
+    def clean(self):
+        """
+        To clean the test environment, including restore VAR file
+        and close the session
+        """
+        if self.session:
+            self.session.close()
+
+
+def run(test, params, env):
+    """
+    Uefishell test:
+    1. Boot up guest from uefishell.iso, to enter uefishell
+    2. Execute uefishell command
+    3. Justify if the response of the commmand correct
+    4. Execute post action out of uefishell
+    5. Clean test environment
+
+    :param test: Kvm test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    uefishell_test = UEFIShellTest(test, params, env)
+    under_fs0 = params.get("under_fs0", "yes")
+    uefishell_test.setup(under_fs0)
+    test_scenarios = params["test_scenarios"]
+    for scenario in test_scenarios.split():
+        command = params["command_%s" % scenario]
+        check_result = params.get("check_result_%s" % scenario)
+        uefishell_test.send_command(command, check_result)
```

Comment:
Because some cases may not just run one command, it is more flexible to define a method for every case than just send_command here. Just like what you do for post_test.

Reply:
@YanhuiMa It is designed for multiple command, `test_scenarios` is a list, separated by a blank space. See line 142 `for scenario in test_scenarios.split():`